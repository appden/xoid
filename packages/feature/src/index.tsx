type Config<T, U> = {
  id?: string | symbol
  defaultOptions?: T
  options?: U
}

export type From = {
  <U extends new (...args: any) => any>(item: U): InstanceType<U>
  <U extends { id: any; type: any }>(item: U): U['type']
}

export abstract class FeatureAbstract<Options, DefaultOptions> {
  externalOptions!: Options
  options!: DefaultOptions & Options
  from!: From
  main?: () => void
}

export type FeatureConstructor = new (options: any) => FeatureAbstract<any, any>

type PrefilledFeature<T> = T extends new (options: any) => infer Q
  ? new (options: {}) => Omit<Q, 'externalOptions'> & { externalOptions: {} }
  : never

export type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never

type GetKey<I, K extends string | symbol | number> = I extends {
  [L in K]: infer P
}
  ? P
  : never

export type ExtractKey<U, K extends string | number | symbol> = GetKey<UnionToIntersection<U>, K>

type MergedTypes<P extends FeatureConstructor> = {
  [K in keyof UnionToIntersection<InstanceType<P>>]: ExtractKey<InstanceType<P>, K>
}

const REFERENCE = Symbol()

const defaultFrom: any = () => console.warn('Please use it inside compose')

const current = { from: defaultFrom }

const createFrom = (contextMap: any) =>
  function (this: any, input: any) {
    const output = contextMap.get(typeof input === 'object' ? input.id : input)
    if (!output) {
      console.error(`Dependency '${input.name}' not found in '${this.constructor.name}'.`)
    }
    return output
  }

const interceptFrom = <T,>(nextFrom: any, fn: () => T) => {
  current.from = nextFrom
  const ans = fn()
  current.from = defaultFrom
  return ans
}

export function Feature<T extends Config<any, any> = { defaultOptions: {}; options: {} }>(
  config = {
    id: Symbol(),
    defaultOptions: {},
    options: {},
  } as T
) {
  return class {
    static id = config.id

    externalOptions!: T['options']
    options!: T['options'] & T['defaultOptions']
    from!: From

    constructor(options: T['options']) {
      this.from = (a: any) => current.from(a)
      this.options = { ...config.defaultOptions, ...options }
    }
  }
}

export const composeFeatures =
  <P extends FeatureConstructor, V>(
    config: P[],
    afterAll: (from: From, common: MergedTypes<P>) => V
  ) =>
  (options: ExtractKey<InstanceType<P>, 'externalOptions'>) => {
    const contextMap = new Map<P, InstanceType<P>>()
    config
      .map((item) => {
        const instance = runInner(contextMap)(item, options)
        // @ts-ignore
        contextMap.set(item[REFERENCE] || item, instance)
        // @ts-ignore
        if (item.id) {
          // @ts-ignore
          contextMap.set(item.id, instance)
        }
        return instance
      })
      .forEach((instance) => {
        instance?.main?.()
      })
    const ans = afterAll(contextMap.get.bind(contextMap) as From, {} as MergedTypes<P>)
    return ans
  }

const runInner =
  (contextMap: Map<FeatureConstructor, FeatureAbstract<any, any>>) =>
  <P extends FeatureConstructor>(
    item: P,
    options: ExtractKey<InstanceType<P>, 'externalOptions'>
  ) => {
    const from = createFrom(contextMap)
    const instance = interceptFrom(from, () => new item(options))
    instance.from = from
    return instance
  }

export const prefill = <
  T extends FeatureConstructor,
  U extends ExtractKey<InstanceType<T>, 'externalOptions'>
>(
  item: T,
  options: U | ((from: From) => U)
) =>
  // @ts-ignore
  class extends item {
    static [REFERENCE] = item
    constructor() {
      // @ts-ignore
      super(typeof options === 'function' ? options(current.from) : options)
    }
  } as PrefilledFeature<T>
