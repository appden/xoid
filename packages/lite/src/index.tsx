export { create } from '../../xoid/src/internal/lite'
