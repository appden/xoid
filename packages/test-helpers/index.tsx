export const debug = (store: any) => {
  const v = {
    self: store,
    selfSerialized: JSON.stringify(store),
    selfTypeof: typeof store,
    selfKeys: isNonPrimitiveStore(store)
      ? Object.keys(store)
      : '[[non-object]]',
    selfSpread: isNonPrimitiveStore(store) ? { ...store } : '[[non-object]]',
    get: store(),
    getSerialized: JSON.stringify(store()),
    getTypeof: typeof store(),
  }
  return v
}

const isNonPrimitiveStore = (store: any) => {
  const value = store()
  return (
    typeof store === 'function' &&
    (typeof value === 'object' || typeof value === 'function')
  )
}
