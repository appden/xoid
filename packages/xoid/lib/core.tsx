import {
  createNotifier,
  createSelector,
  createTarget,
  META,
  USEABLE,
  Init,
  Atom,
} from '@xoid/engine'
import { select } from './lens'

const useable = Symbol()
export type Useable<U> = { [useable]: U }

/**
 * Gets the "useable" of an atom.
 * @see [xoid.dev/docs/api/use](https://xoid.dev/docs/api/use)
 */
export function use<T extends unknown, U>(atom: Atom<T>, fn: (state: T) => U): Atom<U>
export function use<T extends unknown, U extends keyof T>(atom: Atom<T>, fn: U): Atom<T[U]>
export function use<T extends any>(atom: Useable<T>): T
export function use(atom: any, fn?: any): any {
  if (!fn) return (atom as any)[USEABLE]
  return select(atom, fn)
}

/**
 * Creates an atom with the first argument as the initial state.
 * Second argument can be used too attach "useable"s to the atom.
 * @see [xoid.dev/docs/api/create](https://xoid.dev/docs/api/create)
 */

export function create<T>(): Atom<T | undefined>
export function create<T>(init: Init<T>): Atom<T>
export function create<T, U>(init: Init<T>, useable?: (atom: Atom<T>) => U): Atom<T> & Useable<U>
export function create<T, U = undefined>(init?: Init<T>, useable?: (atom: Atom<T>) => U): Atom<T> {
  const meta = { notifier: createNotifier(), node: init }
  const target = createTarget(meta)
  if (typeof init === 'function') createSelector(target as unknown as Atom<T>, init as Function)
  ;(target as any)[META] = meta
  if (useable && typeof useable === 'function') (target as any)[USEABLE] = useable(target as any)
  return target as any
}
