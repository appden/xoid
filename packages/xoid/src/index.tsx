import type { Atom, Stream, Init, GetState, Usable } from './types'
import { createSelector, createApi, INTERNAL } from './internal/utils'
import { createEvent, createInternal } from './internal/lite'

export * from './types'

/**
 * Creates an atom with the first argument as the initial state.
 * Second argument can be used to attach "usables" to the atom.
 * @see [xoid.dev/docs/api/create](https://xoid.dev/docs/quick-tutorial)
 */
export function create<T>(): Stream<T>
export function create<T>(init: Init<T>): Atom<T>
export function create<T>(init: Init<T>, getUsable: null): Atom<T>
export function create<T, U>(init: Init<T>, getUsable?: (atom: Atom<T>) => U): Atom<T> & Usable<U>
export function create<T, U = undefined>(
  init?: Init<T>,
  getUsable?: null | ((atom: Atom<T>) => U)
) {
  const isFunction = typeof init === 'function'
  const initialValue = (isFunction ? undefined : init) as T
  const internal = createInternal(initialValue, (() => (use as any).devtools.send(atom)) as any)
  internal.isStream = !arguments.length
  if (isFunction) internal.get = createSelector(internal, init as (get: GetState) => T)
  const atom = createApi(internal)
  use.plugins.forEach((fn: any) => fn(atom, initialValue))
  internal.usable = getUsable?.(atom)
  return atom
}

/**
 * Gets the "usables" of an atom.
 * @see [xoid.dev/docs/api/use](https://xoid.dev/docs/quick-tutorial)
 */
export const use = <T extends any>(atom: Usable<T>): T =>
  _use.devtools.wrap((atom as any)[INTERNAL].usable, atom)

// Untyped stuff for devtools
const _use = use as any
_use.symbol = INTERNAL
_use.createEvent = createEvent
_use.devtools = {
  send: <T,>(_atom: T) => void 0,
  wrap: <T,>(value: T): T => value,
}

use.plugins = [] as ((atom: Atom<any>, initialValue?: any) => void)[]
