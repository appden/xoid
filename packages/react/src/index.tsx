import { useSyncExternalStore } from 'use-sync-external-store/shim'
import React, { useEffect, useLayoutEffect, useRef, useDebugValue } from 'react'
import { Atom, Usable, create, use as _use } from 'xoid'

export type Pair<T> = { value: T; onChange: (value: T) => void }

export type ComponentType = keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>

export type PropsOf<T> = T extends ComponentType
  ? React.ComponentProps<T>
  : T extends (...args: any) => any
  ? Parameters<T>[0]
  : never

// For server-side rendering: https://github.com/react-spring/zustand/pull/34
export const useIsoLayoutEffect = window === undefined ? useEffect : useLayoutEffect

const useConstant = <T extends any>(fn: () => T): T => {
  const ref = useRef<{ c: T }>()
  if (!ref.current) ref.current = { c: fn() }
  return ref.current.c
}

const useReactAdapter = () => {
  const result = useConstant(() => {
    const mount = (_use as any).createEvent()
    const unmount = (_use as any).createEvent()
    return {
      mount,
      unmount,
      adapter: {
        mount: mount.add,
        unmount: unmount.add,
      },
    }
  })
  useIsoLayoutEffect(() => {
    result.mount.fire()
    return result.unmount.fire
  }, [])
  return result.adapter
}

export type ReactAdapter = ReturnType<typeof useReactAdapter>

/**
 * Subscribes to an atom inside a React function component.
 *
 * An atom, or a function returning an atom can be passed as the first argument.
 * When the second optional argument is set to `true`, it will also consume the usables of the atom.
 * @see [xoid.dev/docs/api-react/use-atom](https://xoid.dev/docs/api-react/use-atom)
 */
export function useAtom<T>(atom: Atom<T>): T
export function useAtom<T>(atom: () => Atom<T>): T
export function useAtom<T, U>(atom: Atom<T> & Usable<U>, use: true): [T, U]
export function useAtom<T, U>(atom: () => Atom<T> & Usable<U>, use: true): [T, U]
export function useAtom<T, U>(maybeAtom: Atom<T> | (() => Atom<T>), use?: boolean): [T, U] | T {
  const atom = useConstant(() => (typeof maybeAtom === 'function' ? maybeAtom() : maybeAtom))
  const value = useSyncExternalStore(
    atom.subscribe,
    () => atom.value,
    () => atom.value
  )
  useDebugValue(value)
  return use ? ([value, _use(atom as any)] as [T, U]) : (value as T)
}

/**
 * Can be used to create local state inside React components. Similar to `React.useMemo`,
 * but creates values **exactly once**.
 * @see [xoid.dev/docs/api-react/use-setup](https://xoid.dev/docs/api-react/use-setup)
 */
export function useSetup<T>(fn: ($props: undefined, adapter: ReactAdapter) => T): T
export function useSetup<T, P>(fn: ($props: Atom<P>, adapter: ReactAdapter) => T, props: P): T
export function useSetup(fn: ($props: any, adapter: any) => any, props?: any): any {
  /* eslint-disable react-hooks/rules-of-hooks */
  // Calling hooks conditionally wouldn't be an issue here, because we rely on just
  // the Function.length and Arguments.length, which will remain static.
  const api = fn.length > 1 ? useReactAdapter() : undefined
  if (arguments.length > 1) {
    const [$props, result] = useConstant(() => {
      const $props = create(() => props)
      const result = fn($props, api)
      return [$props, result] as const
    })
    useIsoLayoutEffect(() => ($props as Atom<any>).set(props), [props])
    return result
  } else {
    const result = useConstant(() => fn(undefined, api))
    return result
  }
  /* eslint-enable react-hooks/rules-of-hooks */
}
