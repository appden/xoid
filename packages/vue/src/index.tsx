import { getCurrentScope, onScopeDispose, shallowRef } from 'vue'
import { Atom } from 'xoid'

export function useAtom<T>(atom: Atom<T>) {
  const state = shallowRef()

  const unsubscribe = atom.subscribe((value) => {
    state.value = value
  })

  getCurrentScope() && onScopeDispose(unsubscribe)
  return state
}
