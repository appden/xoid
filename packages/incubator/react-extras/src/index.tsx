import { useAtom, useSetup } from '@xoid/react'
export { slice } from './slice'

/**
 * Can be used to consume a **value-onChange pair** as an atom inside React components.
 *
 * Third argument, when set to `true`, enables optimistic updates. This can be
 * used for keeping a synced local state, and debouncing the updates sent to the parent.
 * @see [xoid.dev/docs/api-react/use-thru](https://xoid.dev/docs/api-react/use-thru)
 */
export const useThru = <T,>(value: T, onChange?: (value: T) => void, optimistic = false) => {
  const atom = useSetup(
    ($props) => {
      const atom = $props.focus('value')
      const props = $props.value
      const prevSet = atom.set
      atom.set = (value: T) => {
        if (props.optimistic) prevSet(value)
        props.onChange?.(value)
      }
      return atom
    },
    { value, onChange, optimistic }
  )
  useAtom(atom)
  return atom
}
