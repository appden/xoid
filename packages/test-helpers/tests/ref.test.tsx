import { ref } from '../src/core'
import { debug } from '../src/test-helpers'

it('creates a ref with a primitive value', () => {
  const store = ref(5)
  expect(debug(store)).toMatchSnapshot()
})

it.only('creates a ref with an object', () => {
  const store = ref({ val: 25 })
  expect(store.val).toBeUndefined()
  expect(debug(store)).toMatchSnapshot()
})
