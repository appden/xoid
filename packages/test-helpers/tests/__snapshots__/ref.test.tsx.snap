// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a ref with an object 1`] = `
Object {
  "get": bound Object {
    "val": 25,
  },
  "getSerialized": "{\\"val\\":25}",
  "getTypeof": "object",
  "self": [Function],
  "selfKeys": Array [
    "val",
  ],
  "selfSerialized": undefined,
  "selfSpread": Object {
    "val": [Function],
  },
  "selfTypeof": "function",
}
`;
